### PR TITLE
adds three_way_comparable[_with] and enables building with GCC

### PR DESCRIPTION
**Pull request by cjdb**
* adds implementation and tests for `three_way_comparable[_with]`
* reorganises code to depend on `partially-ordered-with`
* enables building with GCC to facilitate testing these changes

### CODE REVIEW

File: test/unit/concepts/comparison/three_way_comparable.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Christopher Di Bella.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#include "cjdb/concepts/comparison/three_way_comparable.hpp"
+
+#include <version>
+#if defined(__cpp_lib_three_way_comparison) and __cpp_lib_three_way_comparison >= 201711
```
Comment (brevzin) [inline test/unit/concepts/comparison/three_way_comparable.cpp:7]: Value should be 201907 (which includes P1614, which adds these concepts, and P1902 which fixes the name of the macro that paper added)

File: test/unit/concepts/comparison/three_way_comparable.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+#include "cjdb/concepts/core/convertible_to.hpp"
+#include "three_way_comparable.hpp"
+
+template<typename T, typename Cat, bool expected>
+constexpr void check_three_way_comparable_impl() noexcept
+{
+	static_assert(cjdb::three_way_comparable<T, Cat> == expected);
+}
+
+template<typename T, cjdb::common_reference_with<T> U, typename Cat, bool expected>
```
Comment (brevzin) [inline test/unit/concepts/comparison/three_way_comparable.cpp:19]: I feel like you just want to do `typename U`, then you can check for cases where `T` and `U` don't have a common reference and that that is `false`.

File: test/unit/concepts/comparison/three_way_comparable.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	check_three_way_comparable_with_impl<T, T, std::weak_ordering,
+	                                cjdb::convertible_to<Cat, std::weak_ordering>>();
+
+	check_three_way_comparable_impl<T, std::strong_ordering,
+	                                cjdb::convertible_to<Cat, std::strong_ordering>>();
+	check_three_way_comparable_with_impl<T, T, std::strong_ordering,
+	                                cjdb::convertible_to<Cat, std::strong_ordering>>();
+
+	check_three_way_comparable_impl<T, int, false>();
+	check_three_way_comparable_with_impl<T, T, int, false>();
```
Comment (brevzin) [inline test/unit/concepts/comparison/three_way_comparable.cpp:42]: We really need named parameters ...

File: test/unit/concepts/comparison/three_way_comparable.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+constexpr void check_three_way_comparable_impl() noexcept
+{
+	static_assert(cjdb::three_way_comparable<T, Cat> == expected);
+}
+
+template<typename T, cjdb::common_reference_with<T> U, typename Cat, bool expected>
+constexpr void check_three_way_comparable_with_impl() noexcept
+{
+	static_assert(cjdb::three_way_comparable_with<T, U, Cat> == expected);
+}
```
Comment (brevzin) [inline test/unit/concepts/comparison/three_way_comparable.cpp:23]: Also like... do you need these two functions? They're _barely_ shorter (and not shorter if you drop the `== true` and replace `== false` with `!`) and it's easier to follow if you just use the named concepts:

```cpp
check_three_way_comparable_impl<T, std::partial_ordering, true>();
static_assert(three_way_comparable<T, std::partial_ordering> == true);
static_assert(three_way_comparable<T, std::partial_ordering>);

check_three_way_comparable_with_impl<T, T, std::partial_ordering, true>();
static_assert(three_way_comparable_with<T, T, std::partial_ordering> == true);
static_assert(three_way_comparable_with<T, T, std::partial_ordering>);
```

Comment (cjdb) [reply test/unit/concepts/comparison/three_way_comparable.cpp:23]: Yeah, good point. These functions are a relic of an older test, where I checked cv-qualified and ref-qualified stuff too.

File: test/unit/concepts/comparison/three_way_comparable.cpp (hunk 1 above)
Comment (cjdb) [reply test/unit/concepts/comparison/three_way_comparable.cpp:7]: Applied in cf7734b02373a2a2bfbfaaa3753692cb89aad6ac.

File: test/unit/concepts/comparison/three_way_comparable.cpp (hunk 2 above)
Comment (cjdb) [reply test/unit/concepts/comparison/three_way_comparable.cpp:19]: This particular test should only permit things with a common reference through, so I think it should remain. The case where there isn't a common reference was missing, and has now been added.

File: test/unit/concepts/comparison/three_way_comparable.cpp (hunk 4 above)
Comment (cjdb) [reply test/unit/concepts/comparison/three_way_comparable.cpp:23]: Applied in cf7734b02373a2a2bfbfaaa3753692cb89aad6ac.

**Pull request merged**